After a Gradle build `java.lang.OutOfMemoryError: Java heap space` crash, show error with instructions for bumping Gradle memory properties

### Use case

Example Java heap space failure for `:app:mergeDebugJavaResource`  (in verbose mode) 
```
 [        ]   Caching not enabled.
 [        ] > Task :app:mergeDebugJavaResource FAILED
 [        ] FAILURE: Build failed with an exception.
 [        ] * What went wrong:
 [        ] Execution failed for task ':app:mergeDebugJavaResource'.
 [        ] > Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
 [        ]    > Failed to transform arm64_v8a_debug-1.0.0-78e5d7995cdb16e0b8a3eb12bb62ff325e8c601d.jar (io.flutter:arm64_v8a_debug:1.0.0-78e5d7995cdb16e0b8a3eb12bb62ff325e8c601d) to match attributes {artifactType=android-java-res, org.gradle.category=library, org.gradle.libraryelements=jar, org.gradle.status=release, org.gradle.usage=java-runtime}.
 [        ]       > Execution failed for JetifyTransform: C:\b\s\w\ir\cache\gradle\caches\modules-2\files-2.1\io.flutter\arm64_v8a_debug\1.0.0-78e5d7995cdb16e0b8a3eb12bb62ff325e8c601d\26e409deb8a510a12d4546e73f22c9ecddb96f3\arm64_v8a_debug-1.0.0-78e5d7995cdb16e0b8a3eb12bb62ff325e8c601d.jar.
 [        ]          > Java heap space
```
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8723297509138408161/+/u/run_test.dart_for_build_tests_shard_and_subshard_2_9/stdout




### Proposal

Instead, show a nice error that suggest the user update their `gradle.properties` `org.gradle.jvmargs`.

--- a/packages/flutter_tools/lib/src/android/gradle_errors.dart
+++ b/packages/flutter_tools/lib/src/android/gradle_errors.dart
@@ -352,19 +352,13 @@ final transformInputIssueHandler = GradleHandledError(
 final javaHeapSpaceHandler = GradleHandledError(
   test: _lineMatcher(const <String>['Java heap space']),
   handler: ({required String line, required FlutterProject project, required bool usesAndroidX}) async {
-    final File gradlePropertiesFile = project.android.hostAppGradleRoot.childFile('gradle.properties');
     final String textInBold = globals.logger.terminal.bolden(
-      'Open ${gradlePropertiesFile.path} and ensure org.gradle.jvmargs includes enough heap.\n'
-      'If org.gradle.jvmargs is already set, append these flags to the existing value '
-      '(keep your other JVM arguments, such as file encoding):\n'
-      '-Xmx4G -XX:MaxMetaspaceSize=2G\n'
-      'If org.gradle.jvmargs is not set, add a line such as:\n'
-      'org.gradle.jvmargs=-Xmx4G -XX:MaxMetaspaceSize=2G',
+      'Adjust the maximum Java heap allocation according to the documentation:\n'
+      'https://docs.gradle.org/current/userguide/config_gradle.html#sec:configuring_jvm_memory',
     );
     globals.printBox(
       '${globals.logger.terminal.warningMark} The Gradle build ran out of Java heap space.\n'
-      '$textInBold\n\n'
-      'If this error continues, increase the -Xmx and -XX:MaxMetaspaceSize values based on your machine resources.',
+      '$textInBold',
       title: _boxTitle,
     );
     return GradleBuildStatus.exit;

--- a/packages/flutter_tools/lib/src/android/gradle_errors.dart
+++ b/packages/flutter_tools/lib/src/android/gradle_errors.dart
@@ -70,6 +70,7 @@ final gradleErrors = <GradleHandledError>[
   r8DexingBugInAgp73Handler,
   minSdkVersionHandler,
   transformInputIssueHandler,
+  javaHeapSpaceHandler,
   lockFileDepMissingHandler,
   minCompileSdkVersionHandler,
   incompatibleJavaAndAgpVersionsHandler,
@@ -344,6 +345,27 @@ final transformInputIssueHandler = GradleHandledError(
         return GradleBuildStatus.exit;
       },
   eventLabel: 'transform-input-issue',
+);
+
+/// Handler when a Gradle task fails due to Java heap space exhaustion.
+@visibleForTesting
+final javaHeapSpaceHandler = GradleHandledError(
+  test: _lineMatcher(const <String>['Java heap space']),
+  handler: ({required String line, required FlutterProject project, required bool usesAndroidX}) async {
+    final File gradlePropertiesFile = project.android.hostAppGradleRoot.childFile('gradle.properties');
+    final String textInBold = globals.logger.terminal.bolden(
+      'Open ${gradlePropertiesFile.path} and update (or add) this line:\n'
+      'org.gradle.jvmargs=-Xmx4G -XX:MaxMetaspaceSize=2G',
+    );
+    globals.printBox(
+      '${globals.logger.terminal.warningMark} The Gradle build ran out of Java heap space.\n'
+      '$textInBold\n\n'
+      'If this error continues, increase these values further based on your machine resources.',
+      title: _boxTitle,
+    );
+    return GradleBuildStatus.exit;
+  },
+  eventLabel: 'java-heap-space',
 );
 
 /// Handler when a dependency is missing in the lockfile.

--- a/packages/flutter_tools/lib/src/android/gradle_errors.dart
+++ b/packages/flutter_tools/lib/src/android/gradle_errors.dart
@@ -354,13 +354,17 @@ final javaHeapSpaceHandler = GradleHandledError(
   handler: ({required String line, required FlutterProject project, required bool usesAndroidX}) async {
     final File gradlePropertiesFile = project.android.hostAppGradleRoot.childFile('gradle.properties');
     final String textInBold = globals.logger.terminal.bolden(
-      'Open ${gradlePropertiesFile.path} and update (or add) this line:\n'
+      'Open ${gradlePropertiesFile.path} and ensure org.gradle.jvmargs includes enough heap.\n'
+      'If org.gradle.jvmargs is already set, append these flags to the existing value '
+      '(keep your other JVM arguments, such as file encoding):\n'
+      '-Xmx4G -XX:MaxMetaspaceSize=2G\n'
+      'If org.gradle.jvmargs is not set, add a line such as:\n'
       'org.gradle.jvmargs=-Xmx4G -XX:MaxMetaspaceSize=2G',
     );
     globals.printBox(
       '${globals.logger.terminal.warningMark} The Gradle build ran out of Java heap space.\n'
       '$textInBold\n\n'
-      'If this error continues, increase these values further based on your machine resources.',
+      'If this error continues, increase the -Xmx and -XX:MaxMetaspaceSize values based on your machine resources.',
       title: _boxTitle,
     );
     return GradleBuildStatus.exit;

--- a/packages/flutter_tools/lib/src/android/gradle_errors.dart
+++ b/packages/flutter_tools/lib/src/android/gradle_errors.dart
@@ -351,18 +351,19 @@ final transformInputIssueHandler = GradleHandledError(
 @visibleForTesting
 final javaHeapSpaceHandler = GradleHandledError(
   test: _lineMatcher(const <String>['Java heap space']),
-  handler: ({required String line, required FlutterProject project, required bool usesAndroidX}) async {
-    final String textInBold = globals.logger.terminal.bolden(
-      'Adjust the maximum Java heap allocation according to the documentation:\n'
-      'https://docs.gradle.org/current/userguide/config_gradle.html#sec:configuring_jvm_memory',
-    );
-    globals.printBox(
-      '${globals.logger.terminal.warningMark} The Gradle build ran out of Java heap space.\n'
-      '$textInBold',
-      title: _boxTitle,
-    );
-    return GradleBuildStatus.exit;
-  },
+  handler:
+      ({required String line, required FlutterProject project, required bool usesAndroidX}) async {
+        final String textInBold = globals.logger.terminal.bolden(
+          'Adjust the maximum Java heap allocation according to the documentation:\n'
+          'https://docs.gradle.org/current/userguide/config_gradle.html#sec:configuring_jvm_memory',
+        );
+        globals.printBox(
+          '${globals.logger.terminal.warningMark} The Gradle build ran out of Java heap space.\n'
+          '$textInBold',
+          title: _boxTitle,
+        );
+        return GradleBuildStatus.exit;
+      },
   eventLabel: 'java-heap-space',
 );
 

--- a/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
@@ -42,6 +42,7 @@ void main() {
           r8DexingBugInAgp73Handler,
           minSdkVersionHandler,
           transformInputIssueHandler,
+          javaHeapSpaceHandler,
           lockFileDepMissingHandler,
           minCompileSdkVersionHandler,
           incompatibleJavaAndAgpVersionsHandler,
@@ -806,6 +807,37 @@ assembleProfile
             '└───────────────────────────────────────────────────────────────────────────────┘\n',
           ),
         );
+      },
+      overrides: <Type, Generator>{
+        GradleUtils: () => FakeGradleUtils(),
+        Platform: () => fakePlatform('android'),
+        FileSystem: () => fileSystem,
+        ProcessManager: () => processManager,
+      },
+    );
+  });
+
+  group('java heap space', () {
+    testWithoutContext('pattern', () {
+      expect(javaHeapSpaceHandler.test('> Java heap space'), isTrue);
+      expect(
+        javaHeapSpaceHandler.test('java.lang.OutOfMemoryError: Java heap space'),
+        isTrue,
+      );
+    });
+
+    testUsingContext(
+      'suggestion',
+      () async {
+        final GradleBuildStatus status = await javaHeapSpaceHandler.handler(
+          project: FlutterProject.fromDirectoryTest(fileSystem.currentDirectory),
+          usesAndroidX: true,
+          line: '> Java heap space',
+        );
+
+        expect(status, GradleBuildStatus.exit);
+        expect(testLogger.statusText, contains('org.gradle.jvmargs'));
+        expect(testLogger.statusText, contains('gradle.properties'));
       },
       overrides: <Type, Generator>{
         GradleUtils: () => FakeGradleUtils(),

--- a/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
@@ -836,8 +836,11 @@ assembleProfile
         );
 
         expect(status, GradleBuildStatus.exit);
-        expect(testLogger.statusText, contains('org.gradle.jvmargs'));
-        expect(testLogger.statusText, contains('gradle.properties'));
+        expect(testLogger.statusText, contains('Java heap space'));
+        expect(
+          testLogger.statusText,
+          contains('https://docs.gradle.org/current/userguide/config_gradle.html#sec:configuring_jvm_memory'),
+        );
       },
       overrides: <Type, Generator>{
         GradleUtils: () => FakeGradleUtils(),

--- a/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
@@ -820,10 +820,7 @@ assembleProfile
   group('java heap space', () {
     testWithoutContext('pattern', () {
       expect(javaHeapSpaceHandler.test('> Java heap space'), isTrue);
-      expect(
-        javaHeapSpaceHandler.test('java.lang.OutOfMemoryError: Java heap space'),
-        isTrue,
-      );
+      expect(javaHeapSpaceHandler.test('java.lang.OutOfMemoryError: Java heap space'), isTrue);
     });
 
     testUsingContext(
@@ -839,7 +836,9 @@ assembleProfile
         expect(testLogger.statusText, contains('Java heap space'));
         expect(
           testLogger.statusText,
-          contains('https://docs.gradle.org/current/userguide/config_gradle.html#sec:configuring_jvm_memory'),
+          contains(
+            'https://docs.gradle.org/current/userguide/config_gradle.html#sec:configuring_jvm_memory',
+          ),
         );
       },
       overrides: <Type, Generator>{

